File: full_case_study_solns/_f_planar_vtol/python/hw2/VTOLSim.py
import matplotlib.pyplot as plt
import numpy as np
import sys
sys.path.append('..')  # add parent directory
import VTOLParam as P
from signalGenerator import signalGenerator
from VTOLAnimation import VTOLAnimation
from dataPlotter import dataPlotter


# instantiate reference input classes
z_reference = signalGenerator(amplitude=0.5, frequency=0.1)
h_reference = signalGenerator(amplitude=0.5, frequency=0.1)
zRef = signalGenerator(amplitude=0.5, frequency=0.1)
hRef = signalGenerator(amplitude=0.5, frequency=0.1)
thetaRef = signalGenerator(amplitude=2.0*np.pi, frequency=0.1)
fRef = signalGenerator(amplitude=5, frequency=0.5)
tauRef = signalGenerator(amplitude=5, frequency=0.5)

# instantiate the simulation plots and animation
dataPlot = dataPlotter()
animation = VTOLAnimation()

t = P.t_start  # time starts at t_start
while t < P.t_end:  # main simulation loop
    # set variables
    z_r = z_reference.sin(t)
    z = zRef.sin(t)
    h_r = h_reference.square(t)
    h = zRef.sin(t)
    theta = thetaRef.sin(t)
    f = fRef.sawtooth(t)
    tau = fRef.sawtooth(t)
    # update animation
    state = np.array([[z], [h], [theta], [0.0], [0.0], [0.0]])
    animation.update(state, z_r)
    dataPlot.update(t, state, z_r, h_r, f, tau)

    t = t + P.t_plot  # advance time by t_plot
    plt.pause(0.1)

# Keeps the program from closing until the user presses a button.
print('Press key to close')
plt.waitforbuttonpress()
plt.close()

File: full_case_study_solns/_f_planar_vtol/matlab_old/hw18/VTOLSim.m
if 1
    % VTOL parameter file
    addpath ./.. % adds the parent directory to the path
    VTOLParam % general parameters
    % load loopshaping design
    loopshape_vtol_lon
    loopshape_vtol_lat_in
    loopshape_vtol_lat_out
end

% instantiate VTOL, controller, and reference input classes 
% Instantiate Dynamics class
VTOL = VTOLDynamics(P);  
ctrl = VTOLController(P);  
z_reference = signalGenerator(4, 0.02);  
h_reference = signalGenerator(3, 0.03);  

% instantiate the data plots and animation
dataPlot = plotData(P);
animation = VTOLAnimation(P);

% main simulation loop
t = P.t_start;  % time starts at t_start
while t < P.t_end  
    % Get referenced inputs from signal generators
    z_ref = 5+z_reference.square(t);
    h_ref = 5+h_reference.square(t);
    % Propagate dynamics in between plot samples
    t_next_plot = t + P.t_plot;
    while t < t_next_plot % updates control and dynamics at faster simulation rate
        u = ctrl.u([z_ref; h_ref], VTOL.outputs());  % Calculate the control value
        VTOL.propagateDynamics(P.mixing*u);  % Propagate the dynamics
        t = t + P.Ts; % advance time by Ts
    end
    % update animation and data plots
    animation.drawVTOL(VTOL.states, 0.0);
    dataPlot.updatePlots(t, VTOL.states, z_ref, h_ref, u(1), u(2));
end


